Clarify parameters for emtting a log record (#3354)
open-telemetry/opentelemetry-specification: specification/logs/bridge-api.md

The `Logger` MUST provide functions to:

- [Emit a `LogRecord`](#emit-a-logrecord) (see the section on `LogRecord`)

> ## LogRecord

Reviewer: This LogRecord section appears to be an orphan, since no other section refers to it.

Author: Removed per https://github.com/open-telemetry/opentelemetry-specification/pull/3354#discussion_r1156539431

`LogRecord`s encapsulate the fields identified in the `LogRecord`
[data model](data-model.md).

> ### Emit a LogRecord

Reviewer: "Emit a LogRecord" is a Logger operation and should be nested under "Logger operations".

Author: Agree it's a strange flow. It mimics what we have in the trace spec.

Tracer refers to span creation [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#tracer) and points to [Create a new Span](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span-creation) which falls under the `Span` section.

Spans are admittedly more complex than LogRecords because Spans have their own operations (e.g., SetStatus, AddEvent, etc), so maybe this is what inspired this flow in the trace spec.

I do not have a strong opinion that we attempt to perfectly mimic the trace spec, however, if we do move this up to the Logger section then I am in support of removing the LogRecord section altogether. Do you agree?

The `Logger` MUST provide functions to:

- Emit a `LogRecord`

#### Emit a LogRecord

> `LogRecord`s encapsulate the fields identified in the `LogRecord`
> [data model](data-model.md) and are emitted to the processing pipeline using

Reviewer 2: ```suggestion
`LogRecord`s encapsulate the fields identified in the [`LogRecord`
data model](data-model.md#log-and-event-record-definition) and are emitted to the processing pipeline using
```

this API.

> The API MUST accept the following parameters:

Reviewer 2: Does this mean "Timestamp" MUST be a separate parameter, or it can be part of compound type (e.g. timestamp is part of LogRecord struct)?

Author: The intent is not to dictate how a language's API implements accepting any of these parameters. Separate parameter or part of compound type are both acceptable options.

- [Timestamp](./data-model.md#field-timestamp)
- [Observed Timestamp](./data-model.md#field-observedtimestamp)
- The [Context](../context/README.md) associated with the `LogRecord`. The API

>   MAY also have an option for implicitly using the current Context as a default

Reviewer 2: This might be interpreted in two ways:
1. There may be an option, and the option is off by default. If the user turned the option on, the "current Context" will be used unless the user explicitly specify a Context in the API.
2. There may be an option, and the option is on by default, which means the "current Context" will be used unless the user explicitly specify a Context in the API. The user can turned the option off if such behavior is undesired.

Author: This language is fork-lifted from the trace specification.

https://github.com/open-telemetry/opentelemetry-specification/blob/d25734a47e6534a1bc8a52cb2bd51d42e987024b/specification/trace/api.md?plain=1#L365-L370

Do you know the history of how folks have interpreted this in the trace spec? Have other languages implemented this as some kind of config option?

My intended interpretation was closer to your option 2 minus "The user can turned the option off if such behavior is undesired".

I'm mostly familiar with .NET. Our default behavior with traces is simply to use the current context - there is no configuration option to adjust this behavior - furthermore, there currently is no API provided for explicitly passing in a different context. My read and intent for adopting these words from the trace spec was to allow languages like .NET to do for logs what it currently does for traces.

Reviewer 2: > Do you know the history of how folks have interpreted this in the trace spec? Have other languages implemented this as some kind of config option?

I don't. I think C++ is using the TLS (thread local storage) as implicit context by default and it allows explicit context to be passed in via API.

Author: Maybe "also have an option for" is confusing language. What if we removed it? If we do, I could fix it in the trace spec in a follow up PR... assuming it would be seen as a clarification and not a breaking change.

```suggestion
  MAY implicitly use the current Context as a default
```

Emit a `LogRecord` to the processing pipeline.

This function MAY be named `emit`.

The API MUST accept the following parameters:

* The [LogRecord](#logrecord) to emit.
* The [Context](../context/README.md) associated with the `LogRecord`. The API
  MAY also have an option for implicitly using the current Context as a default

>   behavior.

Reviewer 3: I am still not quite sure about this. If `emit` accept `Context` then that's where it is supposed to get the `SpanContext` from, right? Then why do we have `TraceContext` listed in the `LogRecord` definition later? What happens if both are specified? Is one supposed to override the other?

Emit a `LogRecord` to the processing pipeline.

This function MAY be named `logRecord`. `LogRecord`s emitted can be

> associated with the [Context](../context/README.md).

Reviewer: If this is the direction we want to go (and I'm fine with it), should `Context` be added as a parameter alongside `logRecord` in the section below? 

I see that the metrics API [uses a strategy similar](https://github.com/open-telemetry/opentelemetry-specification/blame/d25734a47e6534a1bc8a52cb2bd51d42e987024b/specification/metrics/api.md?plain=1#L243-L247) to what you've done here. I find that confusing, but fine with it in the name of consistency.

Author: I do not think it should be added as a parameter alongside `logRecord` as I believe that suggests a specific implementation that is not idiomatic for all languages.

In fact, I'd even be inclined to remove the change to this line altogether as the very spirit of OpenTelemetry's Context is that it should always be available by some means for correlating telemetry signals with each other.

Author: > I see that the metrics API [uses a strategy similar](https://github.com/open-telemetry/opentelemetry-specification/blame/d25734a47e6534a1bc8a52cb2bd51d42e987024b/specification/metrics/api.md?plain=1#L243-L247) to what you've done here.

Reflecting on this some more.. It seems to me the reason the metric API even has this statement is to draw a distinction between sync and async instruments. That is, the purpose of the statement is not actually primarily meant to highlight that context is available for metrics - that is already implied by the Context specification. The log bridge API does not require any special distinctions.

Reviewer: > I do not think it should be added as a parameter alongside logRecord as I believe that suggests a specific implementation that is not idiomatic for all languages.

I don't think anyone is reading these API specs too literally. If they were, the API experiences across languages would look much more similar than they do 😅. If it's a parameter, why not list is under the parameters section? If its not idiomatic to ever explicitly pass context in a particular language, than that language would be deviating from the spec and presumably has a very good reason to do so. 

> In fact, I'd even be inclined to remove the change to this line altogether as the very spirit of OpenTelemetry's Context is that it should always be available by some means for correlating telemetry signals with each other.

That requires too much a priori knowledge for the reader and is easier to misinterpret / get wrong.

Author: From the [Context spec](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/context#overview)

> Users writing instrumentation in languages that use Context implicitly are discouraged from using the Context API directly.

---

> If its not idiomatic to ever explicitly pass context in a particular language, than that language would be deviating from the spec and presumably has a very good reason to do so.

So .NET would not explicitly pass context because it uses Context implicitly. So, if `context` was a parameter then .NET would deviate from the spec in order to avoid deviating from the spec 😆

Reviewer: > Users writing instrumentation in languages that use Context implicitly are discouraged from using the Context API directly.

The users here are consumers of the API. That's a different audience than the authors of the API, which may need to accommodate non-standard situations. Users can comply with this language by not passing the context parameter, if .NET were to provide it. If .NET were not to provide a context parameter, dot net would not support explicit context injection. I don't have context or opinions about whether that's appropriate, but for languages that want to support explicit context injection, or outright require it, context is a parameter like the others.

Author: For the sake of moving forward, I've added Context as a parameter and made this section look more consistent with the span creation spec. I'd argue, if this is the pattern and intent that makes most sense to folk, then perhaps the metric API spec would also benefit from being made consistent.

> This function MAY be named `logRecord`.

Reviewer 4: strange name for a function 🙄 Maybe `emitLogRecord`

Author: Good eye, I just kinda glazed over that. I changed it to just `emit`, but 🤷 happy to change it to whatever folks want (or maybe remove this line? is this typical in the spec?).

Reviewer 2: +1 Emit
